### PR TITLE
refactor: Fix null check and implicit any

### PR DESCRIPTION
**Pull request by xqzlgy2**
This PR enabled TS compiler check for:

- Strict null check
- Implicit any

Also fixed existing errors.

#368

### CODE REVIEW

**Review by jdneo — CHANGES_REQUESTED** [review]

File: src/views/projectNode.ts [hunk 1]
```diff
@@ -69,7 +73,13 @@ export class ProjectNode extends DataNode {
                         if (!rootPackages) {
                             return undefined;
                         }
-                        result = result.concat(...rootPackages);
+                        const packages: INodeData[][] = [];
```
[the remark this answers is not in the document]
Comment (xqzlgy2) [reply src/views/projectNode.ts:76]: currently seems not

File: src/explorerCommands/rename.ts [hunk 2]
```diff
@@ -13,6 +13,10 @@ export async function renameFile(node: DataNode): Promise<void> {
         return;
     }
 
+    if (!node.uri) {
```
Comment (jdneo) [inline src/explorerCommands/rename.ts]: redundant code

File: src/explorerCommands/new.ts [hunk 3]
```diff
@@ -51,10 +51,10 @@ async function getPackageFsPath(node: DataNode): Promise<string> {
         });
         if (packageRoots.length < 1) {
             // This might happen for an invisible project with "_" as its root
-            const packageNode: DataNode = childrenNodes.find((child) => {
+            const packageNode: DataNode | undefined = childrenNodes.find((child) => {
                 return child.nodeData.kind === NodeKind.Package;
             });
-            if (packageNode) {
+            if (packageNode && packageNode.uri) {
```
Comment (jdneo) [inline src/explorerCommands/new.ts:57]: `if (packageNode?.uri)`

File: src/java/hierarchicalPackageNodeData.ts [hunk 4]
```diff
@@ -13,9 +13,9 @@ export class HierarchicalPackageNodeData implements INodeData {
     }
 
     public name: string;
-    public children = [];
+    public children: HierarchicalPackageNodeData[] = [];
     public displayName: string;
-    private nodeData: INodeData = null;
+    private nodeData?: INodeData = undefined;
```
Comment (jdneo) [inline src/java/hierarchicalPackageNodeData.ts:18]: `private nodeData?: INodeData;`

File: src/views/PrimaryTypeNode.ts [hunk 5]
```diff
@@ -49,7 +49,7 @@ export class PrimaryTypeNode extends DataNode {
     }
 
     protected get iconPath(): string | ThemeIcon {
-        switch (this.nodeData.metaData[PrimaryTypeNode.K_TYPE_KIND]) {
+        switch (this.nodeData.metaData![PrimaryTypeNode.K_TYPE_KIND]) {
```
Comment (jdneo) [inline src/views/PrimaryTypeNode.ts:52]: `this.nodeData.metaData?.[PrimaryTypeNode.K_TYPE_KIND]`

File: src/views/hierarchicalPackageRootNode.ts [hunk 6]
```diff
@@ -46,22 +46,22 @@ export class HierarchicalPackageRootNode extends PackageRootNode {
                 }
             });
         }
-        return this.getHierarchicalPackageNodes().concat(result);
+        const hierarchicalPackageNodes = this.getHierarchicalPackageNodes();
```
Comment (jdneo) [inline src/views/hierarchicalPackageRootNode.ts:49]: `return result.concat(hierarchicalPackageNodes)`

**Review by jdneo — CHANGES_REQUESTED** [review]

File: src/views/hierarchicalPackageRootNode.ts [hunk 7]
```diff
@@ -46,22 +46,22 @@ export class HierarchicalPackageRootNode extends PackageRootNode {
...
+        return hierarchicalPackageNodes ? result.concat(hierarchicalPackageNodes) : result;
     }
 
     protected getHierarchicalPackageNodes(): ExplorerNode[] {
         const hierarchicalPackageNodeData = this.getHierarchicalPackageNodeData();
-        return hierarchicalPackageNodeData === null ? [] : hierarchicalPackageNodeData.children.map((hierarchicalChildrenNode) =>
+        return hierarchicalPackageNodeData === undefined ? [] : hierarchicalPackageNodeData.children.map((hierarchicalChildrenNode) =>
```
Comment (jdneo) [inline src/views/hierarchicalPackageRootNode.ts:55]:
```
return hierarchicalPackageNodeData?.children.map(...) || [];
```

File: src/exportJarSteps/ExportJarTaskProvider.ts [hunk 8]
```diff
@@ -188,10 +188,7 @@ class ExportJarTaskTerminal implements Pseudoterminal {
                 const artifactMap: Map<string, string[]> = new Map<string, string[]>();
                 const testOutputFolderMap: Map<string, string[]> = new Map<string, string[]>();
                 const testArtifactMap: Map<string, string[]> = new Map<string, string[]>();
-                const projectList: INodeData[] | undefined = await Jdtls.getProjects(this.stepMetadata.workspaceFolder!.uri.toString());
-                if (!projectList) {
-                    throw new Error("No project found in workspace.");
-                }
+                const projectList: INodeData[] | undefined = await Jdtls.getProjects(this.stepMetadata.workspaceFolder!.uri.toString()) || [];
```
Comment (jdneo) [inline src/exportJarSteps/ExportJarTaskProvider.ts:191]: `!` no need here

**Review by jdneo — CHANGES_REQUESTED** [review]

**Review by jdneo — APPROVED** [review]
Thanks @xqzlgy2

**Pull request merged**